File: app/src/main/java/com/nan/xarch/persistence/XKeyValue.java
package com.nan.xarch.persistence;

import android.content.Context;
import android.os.Parcelable;

import androidx.annotation.NonNull;

import com.alibaba.fastjson.JSON;
import com.nan.xarch.constant.MMKVKey;
import com.tencent.mmkv.MMKV;

import java.util.Map;
import java.util.Set;

/**
 * 本类为MMKV的封装类，防止代码入侵
 */
public final class XKeyValue {

    public static final String TAG = "XMMKV";
    private static final String ID_DEFAULT = "id_default";

    private XKeyValue() {

    }

    public static void init(Context context) {
        MMKV.initialize(context);
    }

    public static MMKV getDefaultMMKV() {
        return MMKV.mmkvWithID(ID_DEFAULT, MMKV.SINGLE_PROCESS_MODE);
    }

    public static <T> void put(@MMKVKey @NonNull String key, @NonNull T value) {
        put(getDefaultMMKV(), key, value);
    }

    public static <T> void put(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull T value) {
        if (value instanceof String || value instanceof Integer || value instanceof Boolean ||
                value instanceof Float || value instanceof Long || value instanceof Double) {
            mmkv.encode(key, String.valueOf(value));
        } else {
            mmkv.encode(key, JSON.toJSONString(value));
        }
    }

    public static void put(@MMKVKey @NonNull String key, @NonNull byte[] value) {
        put(getDefaultMMKV(), key, value);
    }

    public static void put(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull byte[] value) {
        mmkv.encode(key, value);
    }

    public static void put(@MMKVKey @NonNull String key, @NonNull Set<String> value) {
        put(getDefaultMMKV(), key, value);
    }

    public static void put(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull Set<String> value) {
        mmkv.encode(key, value);
    }

    public static void put(@MMKVKey @NonNull String key, @NonNull Parcelable value) {
        put(getDefaultMMKV(), key, value);
    }

    public static void put(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull Parcelable value) {
        mmkv.encode(key, value);
    }

    public static <T> T get(@MMKVKey @NonNull String key, @NonNull T defaultValue) {
        return get(getDefaultMMKV(), key, defaultValue);
    }

    public static <T> T get(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull T defaultValue) {
        String value = mmkv.decodeString(key, String.valueOf(defaultValue));
        if (defaultValue instanceof String) {
            return (T) value;
        } else if (defaultValue instanceof Integer) {
            return (T) Integer.valueOf(value);
        } else if (defaultValue instanceof Boolean) {
            return (T) Boolean.valueOf(value);
        } else if (defaultValue instanceof Float) {
            return (T) Float.valueOf(value);
        } else if (defaultValue instanceof Long) {
            return (T) Long.valueOf(value);
        } else if (defaultValue instanceof Double) {
            return (T) Double.valueOf(value);
        } else {
            return (T) JSON.parseObject(value, defaultValue.getClass());
        }
    }

    public static byte[] get(@MMKVKey @NonNull String key, @NonNull byte[] defaultValue) {
        return get(getDefaultMMKV(), key, defaultValue);
    }

    public static byte[] get(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull byte[] defaultValue) {
        return mmkv.decodeBytes(key, defaultValue);
    }

    public static Set<String> get(@MMKVKey @NonNull String key, @NonNull Set<String> defaultValue) {
        return get(getDefaultMMKV(), key, defaultValue);
    }

    public static Set<String> get(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull Set<String> defaultValue) {
        return mmkv.decodeStringSet(key, defaultValue);
    }

    public static Parcelable get(@MMKVKey @NonNull String key, @NonNull Parcelable defaultValue) {
        return get(getDefaultMMKV(), key, defaultValue);
    }

    public static Parcelable get(MMKV mmkv, @MMKVKey @NonNull String key, @NonNull Parcelable defaultValue) {
        return mmkv.decodeParcelable(key, defaultValue.getClass());
    }

    public static void remove(@MMKVKey @NonNull String key) {
        remove(getDefaultMMKV(), key);
    }

    public static void remove(MMKV mmkv, @MMKVKey @NonNull String key) {
        mmkv.remove(key);
    }

    public static void clear() {
        clear(getDefaultMMKV());
    }

    public static void clear(MMKV mmkv) {
        mmkv.clearAll();
    }

    public static boolean contains(@MMKVKey @NonNull String key) {
        return contains(getDefaultMMKV(), key);
    }

    public static boolean contains(MMKV mmkv, @MMKVKey @NonNull String key) {
        return mmkv.contains(key);
    }

    public static Map<String, ?> getAll() {
        return getAll(getDefaultMMKV());
    }

    public static Map<String, ?> getAll(MMKV mmkv) {
        return mmkv.getAll();
    }

}

File: app/src/main/java/com/nan/xarch/persistence/XKeyValue.kt
package com.nan.xarch.persistence

import android.app.Application
import android.os.Parcelable
import com.nan.xarch.constant.Key
import com.tencent.mmkv.MMKV

/**
 * 本类为MMKV的封装类，防止代码入侵
 */
object XKeyValue {

    private const val ID_DEFAULT = "id_default"
    private const val ID_ACCOUNT = "id_account"

    fun init(application: Application) {
        MMKV.initialize(application)
    }

    fun from(@Key key: String): MMKV {
        return if (key.startsWith("account_")) {
            getAccountMMKV()
        } else {
            getDefaultMMKV()
        }
    }

    private fun getDefaultMMKV(): MMKV {
        return MMKV.mmkvWithID(ID_DEFAULT, MMKV.SINGLE_PROCESS_MODE)
    }

    private fun getAccountMMKV(): MMKV {
        return MMKV.mmkvWithID(ID_ACCOUNT, MMKV.SINGLE_PROCESS_MODE)
    }

    fun clearAccountMMKV() {
        getAccountMMKV().clearAll()
    }

    fun putBoolean(@Key key: String, value: Boolean) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getBoolean(@Key key: String, defaultValue: Boolean = false): Boolean {
        return from(key).decodeBool(key, defaultValue)
    }

    fun putString(@Key key: String, value: String) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getString(@Key key: String, defaultValue: String = ""): String {
        return from(key).decodeString(key, defaultValue)!!
    }

    fun putInt(@Key key: String, value: Int) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getInt(@Key key: String, defaultValue: Int = 0): Int {
        return from(key).decodeInt(key, defaultValue)
    }

    fun putFloat(@Key key: String, value: Float) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getFloat(@Key key: String, defaultValue: Float = 0F): Float {
        return from(key).decodeFloat(key, defaultValue)
    }

    fun putLong(@Key key: String, value: Long) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getLong(@Key key: String, defaultValue: Long = 0L): Long {
        return from(key).decodeLong(key, defaultValue)
    }

    fun putDouble(@Key key: String, value: Double) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getDouble(@Key key: String, defaultValue: Double = 0.0): Double {
        return from(key).decodeDouble(key, defaultValue)
    }

    fun putByteArray(@Key key: String, value: ByteArray) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getByteArray(@Key key: String, defaultValue: ByteArray = ByteArray(0)): ByteArray {
        return from(key).decodeBytes(key, defaultValue)!!
    }

    fun putStringSet(@Key key: String, value: Set<String>) {
        from(key).encode(key, value)
    }

    @JvmOverloads
    fun getStringSet(@Key key: String, defaultValue: Set<String> = mutableSetOf()): Set<String> {
        return from(key).decodeStringSet(key, defaultValue)!!
    }

    fun putParcelable(@Key key: String, value: Parcelable) {
        from(key).encode(key, value)
    }

    inline fun <reified T : Parcelable> getParcelable(@Key key: String): T? {
        return from(key).decodeParcelable(key, T::class.java)
    }
}